### ISSUE TITLE
Fail to understand why the scope name must be specified in module

### ISSUE THREAD

**Issue by feresr** [InsertKoinIO/koin#376]
First of all thanks for putting together such a great library, Its super simple to use and configure, congrats!

I'm trying to understand "scopes" and I fail to see why I need to specify the scope an object will be part of inside the `module`. Like so:

```
val appModule = module {

    // single instance of HelloRepository
    single<HelloRepository> { HelloRepositoryImpl() }

    // Scoped MyScopePresenter instance
    scope("session") { MyScopePresenter(get())}
}
```

Then I have to specify or bind that scope by name again to request the object, Isn't it a bit redundant and restrictive? What's the point?
Why does every object is meant for one scope only? and Why do I need to specify it on the module definition?

TL;DR

I believe modules should be scope agnostic. Koin should figure out what to inject depending which scope the requesting object is in at runtime.

```

// this is completely useless, unncessary IMHO
scope<ActivityA> {}
scope<ActivityB> {}
scope<ActivityC> {}

// Scoped MyScopePresenter instance
scoped { Something(get())}
scoped { Something2(get(), get())}

```

**Comment by feresr** [comment]
Alright, I got around this problem, but I still don't understand why this is designed this way.


Here's how I define my module

```kotlin

val module = module {

    scope<HomeActivity> {}
    scope<CheckoutActivity> {}
    scope<LoginActivity> {}

    scoped { HomeViewModel(get()) }
    scoped { CheckoutViewModel(get()) }
    scoped { LoginViewModel(get()) }
	//other dependencies, they don't care about scope here
}


```

```kotlin

abstract class BaseActivity : AppCompatActivity() {

    override fun onDestroy() {
        if (!isChangingConfigurations) {
            viewModel.onCleared()
            getActivityScope().close()
        }
        super.onDestroy()
    }

	// I HAVE TO DO THIS TO AVOID KOIN TO CREATE NEW SCOPES FOR THE SAME ACTIVITY
	//	fun LifecycleOwner.getScopeName() = this::class.getFullName()
	//	fun LifecycleOwner.getScopeId() = this.toString()
	//	fun getOrCreateAndroidScope(): ScopeInstance {
	//    val name = getScopeName()
	//    val scopeId = getScopeId()
	//   return getKoin().getOrCreateScope(scopeId, name)
	//}
    override fun toString(): String = return this::class.getFullName()
	...
}

```


Then each activity requests what it needs without worring about leaking anything

```kotlin

class MainActivity : BaseActivity() {

override val viewModel: MainViewModel by getActivityScope().inject()

```

**Comment by feresr** [comment]
UPDATE:

After some research I have been able to reduce the boilerplate to this, still... I don't understand why is this so hard to get right.

```kotlin
scope<com.feresr.exposure.Application> {}

scoped { HomeViewModel(get()) }
scoped { CheckoutViewModel(get()) }
scoped { LoginViewModel(get()) }
```

```kotlin
BaseActivity

protected val scope = getKoin().getOrCreateScopeWithType<Application>(this::class.getFullName())
```

**Comment by erickok** [comment]
Are you using Koin 1.x or 2.0 beta? Without knowing _what_ you inject into those view models, it looks like you don't need scoping at all? They can just be factories?

**Comment by feresr** [comment]
I'm using 2.0 beta. 

I use the scope feature to keep the `HomeViewModel` as long as the activity is around. And clear it on

```
    override fun onDestroy() {
        if (!isChangingConfigurations) activityScope.close()
        super.onDestroy()
    }
```

Regardless, I don't get why the scope has to be declaered/named twice?

**Comment by erickok** [comment]
ViewModels are already kept in over configuration changes and properly
released afterwards, you don't have to scope them manually. Just because
that's have Garbage Collection works.

**Comment by feresr** [comment]
I'm not using Android's architecture ViewModels (I have my own reasons for this, one of them is I'm working on a multiplatform project and VM are common to all platforms not android-specific) so I rely on Koin scopes to keep my own viewmodels alive.

Anyways, I got it working and it was super easy to set up with Koin. I just fail to understand why the API ties an object to a specific scope but I'm sure there's a good reason for it. I'll close this now since I got the framework to do what I wanted with my workaround above.

Thanks for reading and answering my questions!
